src/corecel/sys/Stream.cc

```diff
@@ -11,15 +11,50 @@
 #include <iostream>
 
 #include "corecel/Assert.hh"
+#include "corecel/Macros.hh"
 #include "celeritas/Types.hh"
 
 namespace celeritas
 {
+
+//---------------------------------------------------------------------------//
+/*!
+ * Allocate device memory.
+ */
+template<class Pointer>
+auto AsyncMemoryResource<Pointer>::do_allocate(std::size_t bytes, std::size_t)
+    -> pointer
+{
+#if !CELER_USE_DEVICE
+    (void)bytes;
+#endif
+
+    void* ret;
+    CELER_DEVICE_CALL_PREFIX(MallocAsync(&ret, bytes, stream_));
+    return static_cast<pointer>(ret);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Deallocate device memory.
+ */
+template<class Pointer>
+void AsyncMemoryResource<Pointer>::do_deallocate(pointer p,
+                                                 std::size_t,
+                                                 std::size_t)
+{
+#if !CELER_USE_DEVICE
+    (void)p;
+#endif
+
+    CELER_DEVICE_CALL_PREFIX(FreeAsync(p, stream_));
```

Comment:
Since this is presumably called during a destructor, we should probably wrap this in a try/catch that prints the error, as we do in the DeviceAllocation destructor.